daemon/regmaild/regmaild.c

```diff
@@ -516,87 +704,204 @@ regcheck_ambiguous_id(const char *userid)
 //
 // operation: set,count
 
-static void
-err_request(regmaildb_req *req, int fd, struct event *ev)
+static bool
+check_null_terminated(const char *s, size_t sz)
 {
-    fprintf(stderr, "invalid request(%d): uid=[%s]\r\n",
-            req->operation,
-            req->userid);
-    close(fd);
-    free(ev);
+    for (size_t i = 0; i < sz; i++)
+        if (!s[i])
+            return true;
+    return false;
+}
+
+static bool
+validate_regmaildb_request(const regmaildb_req *req)
+{
+    if (req->cb != sizeof(*req)) {
+        fprintf(stderr, "error: corrupted request, cb: %lu != %lu\n",
+                req->cb, sizeof(*req));
+        return false;
+    }
+
+    switch (req->operation)
+    {
+        case REGMAILDB_REQ_COUNT:
+        case REGMAILDB_REQ_SET:
+            if (!*req->userid || !*req->email) {
+                fprintf(stderr, "invalid request: op=[%d], uid=[%s]\n",
+                        req->operation,
+                        req->userid);
+                return false;
+            }
+            break;
+
+        case REGCHECK_REQ_AMBIGUOUS:
+            // No checks.
+            break;
+
+        default:
+            fprintf(stderr, "unknown request: op=[%d]\n", req->operation);
+            return false;
+    }
+    return true;
+}
```

Comment:
This change and other cleanup of regmaildb code could be pulled out into a separate "preparatory" commit.
In a way you will prepare existing code and transform it into a structure that can accept adding the new verifydb
stuff with minimal changes.

Reply:
Separated out.